src/Components/Server/test/Circuits/CircuitPrerendererTest.cs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Server.Circuits;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Components.Server.Tests.Circuits
+{
+    public class CircuitPrerendererTest
+    {
+        // Because CircuitPrerenderer is a point of integration with HttpContext,
+        // it's not a good candidate for unit testing. The majority of prerendering
+        // unit tests should be elsewhere in HtmlRendererTests inside the
+        // Microsoft.AspNetCore.Components.Tests projects.
+        //
+        // The only unit tests added here should specifically be about how we're
+        // interacting with the HttpContext for configuring the prerenderer.
+
+        [Fact]
+        public async Task ExtractsUriFromHttpContext_EmptyPathBase()
+        {
+            // Arrange
+            var circuitFactory = new TestCircuitFactory();
+            var circuitPrerenderer = new CircuitPrerenderer(circuitFactory);
+            var httpContext = new Mock<HttpContext>();
```

Comment:
I strongly dislike mocking (any set of tests based around mocks tends to degrade into a write-only copy-paste fest), so I'm sad to be using it here. But other than handling this entirely as E2E tests, it's probably the only viable way to cover our interaction with `HttpContext`.

Reply:
I forgot to mention. You don't have to mock. You can simply do new DefaultHttpContext() and save yourself some pain. Sorry I missed that.